File: firebase/functions/lib/community/get_members_data.dart
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import 'package:firebase_admin_interop/firebase_admin_interop.dart' as admin;
import '../on_call_function.dart';
import '../utils/firestore_utils.dart';
import '../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/community/member_details.dart';
import 'package:data_models/community/membership.dart';
import 'package:data_models/user/public_user_info.dart';
import 'package:quiver/iterables.dart';

class GetMembersData extends OnCallMethod<GetMembersDataRequest> {
  GetMembersData()
      : super(
          'GetMembersData',
          (jsonMap) => GetMembersDataRequest.fromJson(jsonMap),
        );

  @override
  Future<Map<String, dynamic>> action(
    GetMembersDataRequest request,
    CallableContext context,
  ) async {
    final communityId = request.communityId;

    // Check if current user is admin
    final adminMembershipDocRef =
        'memberships/${context.authUid}/community-membership/$communityId';
    final communityMembershipDoc =
        await firestore.document(adminMembershipDocRef).get();

    final membership = Membership.fromJson(
      firestoreUtils
          .fromFirestoreJson(communityMembershipDoc.data.toMap() ?? {}),
    );

    if (!membership.isAdmin) {
      throw HttpsError(HttpsError.failedPrecondition, 'Unauthorized', null);
    }

    final memberDetailsList = <MemberDetails>[];

    final eventPath = request.eventPath;
    Event? event;
    if (eventPath != null) {
      final eventDoc = await firestore.document(eventPath).get();
      event = Event.fromJson(
        firestoreUtils.fromFirestoreJson(eventDoc.data.toMap()),
      );
    }

    for (var userIdsBatch in partition(request.userIds, 250)) {
      final List<Future<MemberDetails>> memberDetailsListFutures = [];
      for (var member in userIdsBatch) {
        memberDetailsListFutures.add(
          _getDataForUser(
            memberId: member,
            communityId: communityId,
            event: event,
          ),
        );
      }
      memberDetailsList.addAll(await Future.wait(memberDetailsListFutures));
    }

    return GetMembersDataResponse(
      membersDetailsList: memberDetailsList,
    ).toJson();
  }

  Future<MemberDetails> _getDataForUser({
    required String memberId,
    required String communityId,
    required Event? event,
  }) async {
    // Check memberId user membership
    final membershipDocRef =
        'memberships/$memberId/community-membership/$communityId';
    final membershipDoc = await firestore.document(membershipDocRef).get();

    // Get memberId user membership details
    final membershipDetails = membershipDoc.exists
        ? Membership.fromJson(
            firestoreUtils.fromFirestoreJson(membershipDoc.data.toMap() ?? {}),
          )
        : Membership(
            userId: memberId,
            communityId: communityId,
            status: MembershipStatus.nonmember,
          );

    admin.UserRecord? memberInfo;
    try {
      // Get memberId user info
      memberInfo = await firestoreUtils.getUser(memberId);
    } catch (e) {
      print('Error getting user info for $memberId: $e');
    }

    final String memberName;
    if (isNullOrEmpty(memberInfo?.displayName)) {
      final memberDoc = await firestore.document('publicUser/$memberId').get();
      memberName = memberDoc.data.toMap()['displayName'] ?? '';
    } else {
      memberName = memberInfo!.displayName;
    }

    // Get memberId user chats and suggestions per event within current community
    MemberEventData? memberEventsData;

    if (event != null) {
      final String eventPath = event.fullPath;

      final participantData = await firestore
          .document('$eventPath/event-participants/$memberId')
          .get();

      final participant = Participant.fromJson(
        firestoreUtils.fromFirestoreJson(participantData.data.toMap()),
      );

      memberEventsData = MemberEventData(
        eventId: event.id,
        templateId: event.templateId,
        participant: participant,
      );
    }

    return MemberDetails(
      id: memberId,
      email: memberInfo?.email ?? "Unknown",
      displayName: memberName,
      membership: membershipDetails,
      memberEvent: memberEventsData,
    );
  }
}

File: firebase/functions/src/community/get_members_data.ts
import * as functions from 'firebase-functions'
import { HttpsError } from 'firebase-functions/lib/providers/https'
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils'
import { firebaseAuthUtils } from '../utils/infra/firebase_auth_utils'
import { OnCallMethod } from '../on_call_function'
import {
    GetMembersDataRequest,
    GetMembersDataResponse,
    Membership,
    MemberDetails,
    MemberEventData,
    Event,
    Participant,
    ParticipantStatus,
    MembershipStatus,
} from '../types'
import { membershipIsAdmin } from '../types'

export class GetMembersData extends OnCallMethod<GetMembersDataRequest> {
    constructor() {
        super('GetMembersData', (json) => json as GetMembersDataRequest)
    }

    async action(
        request: GetMembersDataRequest,
        context: functions.https.CallableContext
    ): Promise<GetMembersDataResponse> {
        const { communityId } = request

        const adminMembershipDoc = await firestore
            .doc(`memberships/${context.auth?.uid}/community-membership/${communityId}`)
            .get()
        const membership = adminMembershipDoc.data() as Membership | undefined

        if (!membership?.status || !membershipIsAdmin(membership.status)) {
            throw new HttpsError('failed-precondition', 'Unauthorized')
        }

        let event: Event | undefined
        if (request.eventPath) {
            const eventDoc = await firestore.doc(request.eventPath).get()
            event = firestoreUtils.fromFirestoreJson(eventDoc.data() ?? {}) as unknown as Event
            event.id = eventDoc.id
        }

        // Process in batches of 250
        const memberDetailsList: MemberDetails[] = []
        for (let i = 0; i < request.userIds.length; i += 250) {
            const batch = request.userIds.slice(i, i + 250)
            const details = await Promise.all(
                batch.map((memberId) => this._getDataForUser({ memberId, communityId, event }))
            )
            memberDetailsList.push(...details)
        }

        return { membersDetailsList: memberDetailsList }
    }

    private async _getDataForUser(opts: {
        memberId: string
        communityId: string
        event?: Event
    }): Promise<MemberDetails> {
        const { memberId, communityId, event } = opts

        const membershipDoc = await firestore
            .doc(`memberships/${memberId}/community-membership/${communityId}`)
            .get()
        const membershipDetails: Membership = membershipDoc.exists
            ? (membershipDoc.data() as Membership)
            : { userId: memberId, communityId, status: MembershipStatus.nonmember }

        let memberInfo: import('firebase-admin').auth.UserRecord | undefined
        try {
            memberInfo = await firebaseAuthUtils.getUser(memberId)
        } catch (e) {
            console.error(`Error getting user info for ${memberId}:`, e)
        }

        let memberName = ''
        const memberDoc = await firestore.doc(`publicUser/${memberId}`).get()
        if (memberDoc.exists) {
            const publicUserInfo = memberDoc.data() as { displayName?: string }
            memberName = publicUserInfo.displayName ?? ''
        }

        let memberEventsData: MemberEventData | undefined
        if (event) {
            const eventPath = `${event.collectionPath}/${event.id}`
            const participantData = await firestore
                .doc(`${eventPath}/event-participants/${memberId}`)
                .get()
            const participant = participantData.data() as Participant

            memberEventsData = {
                eventId: event.id,
                templateId: event.templateId,
                participant,
            }
        }

        return {
            id: memberId,
            email: memberInfo?.email ?? 'Unknown',
            displayName: memberName,
            membership: membershipDetails,
            memberEvent: memberEventsData,
        }
    }
}
